sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/BufferSendState.scala

```diff
@@ -0,0 +1,232 @@
+          val transferRequest =
+            blockRange.block.bufferTransferRequest
+
+          // we acquire these buffers now, and keep them until the caller releases them
+          // using `releaseAcquiredToCatalog`
+          closeOnExcept(
+            requestHandler.acquireShuffleBuffer(transferRequest.bufferId())) { rapidsBuffer =>
+            //these are closed later, after we synchronize streams
+            rapidsBuffer.storageTier match {
+              case StorageTier.DEVICE =>
+                deviceBuffs += blockRange.rangeSize()
+              case _ => // host/disk
+                hostBuffs += blockRange.rangeSize()
+            }
+            RangeBuffer(blockRange, rapidsBuffer)
+          }
+        }
+
+        logDebug(s"Occupancy for bounce buffer is [device=${deviceBuffs}, host=${hostBuffs}] Bytes")
+
+        bounceBuffToUse = if (deviceBuffs >= hostBuffs || hostBounceBuffer == null) {
+          deviceBounceBuffer.buffer
+        } else {
+          hostBounceBuffer.buffer
+        }
+
+        acquiredBuffs.foreach { case RangeBuffer(blockRange, rapidsBuffer) =>
+          require(blockRange.rangeSize() <= bounceBuffToUse.getLength - buffOffset)
+          withResource(rapidsBuffer.getMemoryBuffer) { memBuff =>
+            bounceBuffToUse match {
+              case _: HostMemoryBuffer =>
+                //TODO: HostMemoryBuffer needs the same functionality that
+                // DeviceMemoryBuffer has to copy from/to device/host buffers
+                CudaUtil.copy(
+                  memBuff,
+                  blockRange.rangeStart,
+                  bounceBuffToUse,
+                  buffOffset,
+                  blockRange.rangeSize())
+              case d: DeviceMemoryBuffer =>
+                memBuff match {
+                  case mh: HostMemoryBuffer =>
+                    // host original => device bounce
+                    d.copyFromHostBufferAsync(buffOffset, mh, blockRange.rangeStart,
+                      blockRange.rangeSize(), serverStream)
+                  case md: DeviceMemoryBuffer =>
+                    // device original => device bounce
+                    d.copyFromDeviceBufferAsync(buffOffset, md, blockRange.rangeStart,
+                      blockRange.rangeSize(), serverStream)
+                  case _ => throw new IllegalStateException("What buffer is this")
+                }
+              case _ => throw new IllegalStateException("What buffer is this")
+            }
+          }
+          buffOffset += blockRange.rangeSize()
+        }
+
+        val alt = AddressLengthTag.from(bounceBuffToUse,
+          blockRanges.head.block.tag)
+
+        alt.resetLength(buffOffset)
+
+        if (windowedBlockIterator.hasNext) {
+          blockRanges = windowedBlockIterator.next()
+        } else {
+          blockRanges = Seq.empty
+          markedDone = true
+        }
+
+        alt
+      } else {
+        null
```

Comment:
Shouldn't this throw `NoSuchElementException` when `next` is called when there are no elements?  Easier to debug than tracking down an NPE that could occur much later.